File: examples/SmartHomeDemo/src/User/SmartHome/RingDoorbell.cs
using System;
using System.Collections.Generic;

namespace SmartHome;

public partial class RingDoorbell
{
    private int _ringCount;
    private string _lastSnapshot = "Garage is calm.";

    public RingDoorbell(string deviceId, string displayName)
    {
        DeviceId = deviceId;
        DisplayName = displayName;
    }

    public HashSet<DeviceFeature> GetSupportedFeatures()
    {
        return
        [
            DeviceFeature.Camera,
            DeviceFeature.Motion,
            DeviceFeature.Doorbell,
            DeviceFeature.Automation,
        ];
    }

    public string DescribeState()
    {
        return $"motion={(MotionDetected ? "detected" : "clear")}, rings={_ringCount}, snapshot=\"{_lastSnapshot}\"";
    }

    public string CaptureSnapshot()
    {
        MotionDetected = false;
        _lastSnapshot = "Driveway is clear and the garage is shut.";
        return _lastSnapshot;
    }

    public void Ring()
    {
        _ringCount++;
        MotionDetected = true;
        _lastSnapshot = "Motion spotted near the garage side entrance.";
    }
}

File: examples/SmartHomeDemo/cpp/user/src/SmartHome/RingDoorbell.cpp
#include "SmartHome/RingDoorbell.hpp"

#include <utility>

namespace SmartHome
{

RingDoorbell::RingDoorbell(::Crimson::Cpp::String deviceId, ::Crimson::Cpp::String displayName)
{
    SetDeviceId(std::move(deviceId));
    SetDisplayName(std::move(displayName));
}

::Crimson::Cpp::Set<DeviceFeature> RingDoorbell::GetSupportedFeatures()
{
    return
    {
        DeviceFeature::Camera,
        DeviceFeature::Motion,
        DeviceFeature::Doorbell,
        DeviceFeature::Automation,
    };
}

::Crimson::Cpp::String RingDoorbell::DescribeState()
{
    return "motion=" + ::Crimson::Cpp::String(GetMotionDetected() ? "detected" : "clear")
        + ", rings=" + std::to_string(ringCount_)
        + ", snapshot=\"" + lastSnapshot_ + "\"";
}

::Crimson::Cpp::String RingDoorbell::CaptureSnapshot()
{
    SetMotionDetected(false);
    lastSnapshot_ = "Driveway is clear and the garage is shut.";
    return lastSnapshot_;
}

void RingDoorbell::Ring()
{
    ++ringCount_;
    SetMotionDetected(true);
    lastSnapshot_ = "Motion spotted near the garage side entrance.";
}

}
